File: Platform/mo-2-core/src/core-java/org/mo/core/aop/config/XAopDefineCollection.java
package org.mo.core.aop.config;

import org.mo.com.lang.FFatalError;
import org.mo.com.lang.RString;

//============================================================
// <T>AOP定义集合。</T>
//============================================================
public class XAopDefineCollection
      extends XAopNodeCollection<XAopDefine>
{
   //============================================================
   // <T>构造AOP定义集合。</T>
   //============================================================
   public XAopDefineCollection(){
      super(XAopDefine.class);
   }

   //============================================================
   // <T>格式化显示内容。</T>
   //============================================================
   public String parse(String value){
      // 检查参数
      if(RString.isEmpty(value)){
         return value;
      }
      // 替换内容
      int start = 0;
      while((start = value.indexOf("${")) != -1){
         int end = value.indexOf("}", start);
         if(-1 != end){
            String refer = value.substring(start + 2, end);
            XAopDefine define = findByKey(refer.toLowerCase());
            if(define == null){
               throw new FFatalError("Refer is not defined. (value={1}, refer={2})", value, refer);
            }
            String parse = define.value();
            if(parse == null){
               parse = "";
            }
            value = value.substring(0, start) + parse + value.substring(end + 1);
         }
      }
      return value;
   }
}

File: Tools/1 - Common/MoCore/Aop/Config/XAopDefineCollection.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Reflection;
using MO.Common.Content;
using MO.Common.Lang;

namespace MO.Core.Aop.Config {

   public class XAopDefineCollection : XAopNodeCollection<XAopDefine> {

      public const string TAG = "Defines";

      public XAopDefineCollection() {
         _duplicate = false;
      }

      public string Parse(string value) {
         if (value != null) {
            while (true) {
               int start = value.LastIndexOf("${");
               if (start == -1) {
                  break;
               }
               int end = value.IndexOf("}", start);
               if (end == -1) {
                  break;
               }
               string name = value.Substring(start + 2, end - start - 2);
               // Find value
               XAopDefine xdefine = _ids[name];
               if (xdefine != null) {
                  value = value.Replace("${" + name + "}", xdefine.Value);
               } else {
                  value = value.Replace("${" + name + "}", "");
               }
            }
         }
         return value;
      }


   }

}
